src/libexpr/primops.cc

```diff
@@ -1152,16 +1152,14 @@ drvName, Bindings * attrs, Value & v)
                 if (i->value->type() == nNull) continue;
             }
 
-            if (i->name == state.sContentAddressed) {
-                contentAddressed = state.forceBool(*i->value, noPos, context_below);
-                if (contentAddressed)
-                    experimentalFeatureSettings.require(Xp::CaDerivations);
+            if (i->name == state.sContentAddressed && state.forceBool(*i->value, noPos, context_below)) {
+                contentAddressed = true;
+                experimentalFeatureSettings.require(Xp::CaDerivations);
             }
 
-            else if (i->name == state.sImpure) {
-                isImpure = state.forceBool(*i->value, noPos, context_below);
-                if (isImpure)
-                    experimentalFeatureSettings.require(Xp::ImpureDerivations);
+            else if (i->name == state.sImpure && state.forceBool(*i->value, noPos, context_below)) {
+                isImpure = true;
+                experimentalFeatureSettings.require(Xp::ImpureDerivations);
```

Comment:
These changes don't look like a change in behavior to me.
For example when `__impure = false;` is passed, then `isImpure` would end up `false` both before and after this PR, iiuc.

Reply:
If I read correctly (I haven't tested this yet), if `__impure = false;`, it will be added to the environment instead of being elided entirely, matching Nix 2.3 behavior. I do agree that the code is very unclear on this :p

Reply:
Ah that is an interesting point. I think the real solution is to find a way to do forwards compat for toggles that don't yet exist. I.e. something that if *not* recognized would cause an error rather than an environment variable entry.

Reply:
In the future we should use `requiredSystemFeatures` for this, e.g. `requiredSystemFeatures = [ "impure" ]`. That triggers an error if you don't have the required feature.

Reply:
Except `requiredSystemFeatures` are only checked on build time, not when instantiating the derivation - so features that modify the way a derivation is made (e.g. `__impure` and `__contentAddressed`) can't work like that.

Have you maybe considered just requiring all derivation attributes starting with `__` to be known to the interpreter?